### PR TITLE
Move internal implementation of the odsp driver cache to the Fluid Framework Repository

### PR DESCRIPTION
**Pull request by christiango**
We have had an internal indexeddb based implementation of the odsp driver cache that our web products have been using for quite some time. Recently, this implementation is being used by partner teams and causing some circular repository dependencies between different products. This PR moves this cache implementation to the open source repo for the following reasons:
* Fix circular repository dependencies
* Enable usage by the community
* Allowing us to fix perf issues with the cache implementation on web in a centralized location.

### CODE REVIEW

**Comment by tylerbutler** [comment]
Who becomes the "owner" of this code (i.e. who should be in the CODEOWNERS file for this new project? @microsoft/fluid-cr-odsp?

I'd prefer not to include ODSP stuff in the core repo but I assume that ship has long since sailed.

**Comment by christiango** [comment]
> Who becomes the "owner" of this code (i.e. who should be in the CODEOWNERS file for this new project)? @microsoft/fluid-cr-odsp?
> 
> I'd prefer not to include ODSP stuff in the core repo but I assume that ship has long since sailed.

Cosmin's team has ben owning it on the FFX side, do you have docs on how you use code owners? I can include the changes in this PR after I fix the lint issues

**Comment by christiango** [comment]
> > Who becomes the "owner" of this code (i.e. who should be in the CODEOWNERS file for this new project)? @microsoft/fluid-cr-odsp?
> > I'd prefer not to include ODSP stuff in the core repo but I assume that ship has long since sailed.
> 
> Cosmin's team has ben owning it on the FFX side, do you have docs on how you use code owners? I can include the changes in this PR after I fix the lint issues

I see this is actually already covered by this rule, so I think we should be ok since that seems like a reasonable group of folks to review PRs here:

/packages/drivers/odsp*/                                                  @microsoft/fluid-cr-drivers @microsoft/fluid-cr-odsp

File: api-report/odsp-driver-web-cache.api.md
Comment (vladsud) [inline api-report/odsp-driver-web-cache.api.md:3]: is there anything ODSP specific for package to have ODSP in name?
I believe R11S is looking to have similar functionality and likely would love to reuse code if it fits their needs.
@znewton - FYI

File: packages/drivers/odsp-driver-web-cache/src/FluidCache.ts
Comment (christiango) [reply packages/drivers/odsp-driver-web-cache/src/FluidCache.ts:159]: I don't want to touch the code during the migration to minimize risk. I had to touch it in a few places to conform with lint rules unfortunately

File: api-report/odsp-driver-web-cache.api.md
Comment (christiango) [reply api-report/odsp-driver-web-cache.api.md:3]: Persisted cache is a contract in ODSP driver, is that moving out?

Comment (christiango) [reply api-report/odsp-driver-web-cache.api.md:3]: You can see all the imports we have from odsp-driver-definitions

File: packages/drivers/odsp-driver-web-cache/src/test/FluidCache.test.ts
Comment (vladsud) [inline packages/drivers/odsp-driver-web-cache/src/test/FluidCache.test.ts:47]: I assume there is no way to run this test (and shipping code in general) in Node environment, as there is no equivalence for indexDb, right? Most of our code is isomorphic, and it would be great if we can use and test this functionality in our stress tests (they are run in node environment)

File: packages/drivers/odsp-driver-web-cache/src/FluidCache.ts
Comment (christiango) [reply packages/drivers/odsp-driver-web-cache/src/FluidCache.ts:159]: But if you think this is a safe change, I would hope so. I can make the change for this case

File: api-report/odsp-driver-web-cache.api.md
Comment (vladsud) [reply api-report/odsp-driver-web-cache.api.md:3]: Got it. I'll leave it up to @znewton to comment here - I believe he was planning to make these interfaces generic (i.e., moving out of ODSP code base)

Comment (christiango) [reply api-report/odsp-driver-web-cache.api.md:3]: But I can rename it driver-web-cache if we confident it eventually will have no ODSP contract dependencies? I have my doubts due to the "file" nature of this

File: packages/drivers/odsp-driver-web-cache/src/FluidCache.ts
Comment (vladsud) [reply packages/drivers/odsp-driver-web-cache/src/FluidCache.ts:159]: Maybe clean it before moving :) After is fine too.

**Review by vladsud — APPROVED** [review]
FYI - I only briefly looked through functionality itself

File: packages/drivers/odsp-driver-web-cache/src/test/FluidCache.test.ts
Comment (christiango) [reply packages/drivers/odsp-driver-web-cache/src/test/FluidCache.test.ts:47]: Yeah this has to run in a browser or mock browser environment. A stress test for this would need to run in a browser context.

File: api-report/odsp-driver-web-cache.api.md
Comment (christiango) [reply api-report/odsp-driver-web-cache.api.md:3]: Ok, I'll rename it!

Comment (christiango) [reply api-report/odsp-driver-web-cache.api.md]: @vladsud - should I wait for @znewton to chime in or just merge with the generic name and rename later if we can't align on the interfaces?

Comment (vladsud) [reply api-report/odsp-driver-web-cache.api.md]: I'd keep all the names generic and merge (even though interfaces are coming from ODSP). I think we will want to move them out of ODSP no matter what.

File: packages/drivers/driver-web-cache/package.json
Comment (heliocliu) [inline packages/drivers/driver-web-cache/package.json:3]:
```suggestion
  "version": "0.58.2000",
```
these were bumped during the course of this review

File: packages/drivers/driver-web-cache/package.json
Comment (heliocliu) [inline packages/drivers/driver-web-cache/package.json:41]:
```suggestion
    "@fluidframework/odsp-driver-definitions": "^0.58.2000",
    "@fluidframework/telemetry-utils": "^0.58.2000",
```

**Pull request merged**
